runtime-sdk/src/context.rs

```diff
@@ -130,75 +120,116 @@ pub trait Context {
     /// consumes the transaction context.
     fn commit(self) -> (Tags, Vec<(roothash::Message, MessageEventHookInvocation)>);
 
-    /// Fetches or sets a value associated with the context.
-    fn value<V>(&mut self, key: &'static str) -> &mut V
-    where
-        V: Any + Default;
+    /// Fetches a value entry associated with the context.
+    fn value<V: Any>(&mut self, key: &'static str) -> ContextValue<'_, V>;
 
-    /// Takes a value associated with the context.
+    /// Executes a function in a simulation context.
     ///
-    /// The previous value is removed so subsequent fetches will return the default value.
-    fn take_value<V>(&mut self, key: &'static str) -> Box<V>
+    /// The simulation context collects its own messages and starts with an empty set of context
+    /// values.
+    fn with_simulation<F, Rs>(&mut self, f: F) -> Rs
     where
-        V: Any + Default;
+        F: FnOnce(
+            RuntimeBatchContext<'_, Self::Runtime, storage::OverlayStore<&mut Self::Store>>,
+        ) -> Rs;
+}
 
-    /// Fetches or sets a value associated with the transaction.
-    ///
-    /// Only present if this is a transaction processing context.
-    fn tx_value<V>(&mut self, key: &'static str) -> Option<&mut V>
+/// Runtime SDK batch-wide context.
+pub trait BatchContext: Context {
```

Comment:
what other things do you envision will impl this and TxContext?

Reply:
I don't think anything will. The trait could be turned into an `impl` if you don't think there'll be mocked versions ever (I think the MKVS is a better mock target, though).

Reply:
It looks to me that that the goal here is to have have a trait (`TxContext`) and struct (`RuntimeTxContext`) be tightly coupled, and then we can use the trait where it fits nicely (static dispatch, generic functions - e.g. all the generic module transaction methods which accept `TxContext`) and the struct in places where we want to avoid dynamic dispatch/trait objects which tend to be somewhat limited/awkward to work with? e.g. bellow in the `with_tx` `FnOnce` argument.
Is this somewhat correct?

Reply:
Actually as @nhynes mentioned we could even get rid of the more specific tx/batch traits and just use concrete types instead. The only reason why you would want traits is if you had different implementations for them which seems unlikely in this case (e.g., you wouldn't mock contexts).

Reply:
Oh but these not being traits makes it much less convenient as then you need to transfer around even more generic arguments that associated types conveniently hide for you.

Reply:
I think what i'm asking is why does the `with_tx` method bellow use `RuntimeTxContext` and not the `TxContext` trait in the `FnOnce`? 
I think it's cause using a trait object would be quite inconvenient, or is there any other reason

Reply:
> I think it's cause using a trait object would be quite inconvenient, or is there any other reason?

Yeah, so a trait object would be impossible due to generic type parameters and closures cannot have generic type parameters AFAIK.